src/extensions/sb-ncbr/tunnels/algorithm.ts

```diff
@@ -0,0 +1,633 @@
+            `${tunnel.props.type} ${tunnel.props.id}` :
+            'Tunnel';
+
+    if (showRadii)
+        return Shape.create(
+            name,
+            tunnel.props,
+            mesh,
+            () => Color(color),
+            () => 1,
+            (i) => `[${processedData[i].X.toFixed(3)}, ${processedData[i].Y.toFixed(3)}, ${processedData[i].Z.toFixed(3)}] - radius: ${processedData[i].Radius.toFixed(3)}`,
+        );
+    return Shape.create(
+        name,
+        tunnel.props,
+        mesh,
+        () => Color(color),
+        () => 1,
+        () => name,
+    );
+}
+
+export async function createTunnelShape(tunnel: Tunnel, color: Color, resolution: number, sampleRate: number, fillFactor: number, webgl: WebGLContext | undefined, prev?: Shape<Mesh>) {
+    let mesh;
+    if (prev) {
+        mesh = await createTunnelMesh(tunnel.data, resolution, sampleRate, fillFactor, webgl, prev.geometry);
+    } else {
+        mesh = await createTunnelMesh(tunnel.data, resolution, sampleRate, fillFactor, webgl);
+    }
+
+
+    const name = tunnel.props.loci ?
+        tunnel.props.loci :
+        tunnel.props.type && tunnel.props.id ?
+            `${tunnel.props.type} ${tunnel.props.id}` :
+            'Tunnel';
+
+    return Shape.create(
+        name,
+        tunnel.props,
+        mesh,
+        () => Color(color),
+        () => 1,
+        () => name,
+    );
+}
+
+function profileToVec3(profile: Profile): Vec3 {
+    return Vec3.create(profile.X, profile.Y, profile.Z);
+}
+
+function calculateCurvature(Pprev: Vec3, Pi: Vec3, Pnext: Vec3) {
+    const V1 = Vec3.sub(Vec3.zero(), Pi, Pprev);
+    const V2 = Vec3.sub(Vec3.zero(), Pnext, Pi);
+    Vec3.normalize(V1, V1);
+    Vec3.normalize(V2, V2);
+    const dotProduct = Vec3.dot(V1, V2); // Cosine of angle between V1 and V2
+    const angle = Math.acos(dotProduct); // Angle in radians
+    return angle; // Higher angle means higher curvature
+}
+
+function determineNumPoints(distance: number, minDistance: number, pointsPerUnitDistance: number) {
+    if (distance <= minDistance) return 1;
+    return Math.ceil(distance * pointsPerUnitDistance);
+}
+
+function shouldInterpolate(Pprev: Vec3, Pi: Vec3, Pnext: Vec3 | undefined, curveThreshold: number, distanceThreshold: number) {
+    const curvature = Pnext ? calculateCurvature(Pprev, Pi, Pnext) : 0;
+    const distance1 = Vec3.distance(Pprev, Pi);
+    const distance2 = Pnext ? Vec3.distance(Pi, Pnext) : 0;
+
+    return curvature > curveThreshold || distance1 > distanceThreshold || distance2 > distanceThreshold;
+}
+
+// Centripetal Catmull–Rom spline interpolation
+function processProfiles(profiles: Profile[], sampleRate: number, fillFactor: number, tension: number = 0.5, numPointsBetween: number = 5): Profile[] {
+    const skipRate = sampleRate;
+    const sampledProfiles = skipRate === 1 ? profiles : profiles.filter((_, index) => index % skipRate === 0 || index === profiles.length - 1); // ensuring the last profile is included
```

Comment:
I don't think this is correct. The idea was for the `sampleRate` to correspond to the number of spheres needed to represent the tunnel, meaning that there should be "1 sphere every 1/sampleRate angstroms".

Skipping profile points from the input only reduces the resolution of this approach and introduces additional errors.